xla/service/hlo_computation_deduplicator_test.cc

```diff
@@ -618,5 +618,35 @@ TEST_F(HloComputationDeduplicatorTest, DontDeduplicateReduceAllReduce) {
   auto computation_names = RunDeduplicatePass(text, /*expect_true=*/false);
   EXPECT_EQ(computation_names.size(), 3);
 }
+
+TEST_F(HloComputationDeduplicatorTest, RemoveDuplicateAllReduce) {
+  const std::string_view text = R"(
+  HloModule TestModule
+
+  add.1 {
+    Arg_0 = s32[] parameter(0)
+    Arg_1 = s32[] parameter(1)
+    ROOT add.2 = s32[] add(Arg_0, Arg_1)
+  }
+  add.2 {
+    Arg_0 = s32[] parameter(0)
+    Arg_1 = s32[] parameter(1)
+    ROOT add.2 = s32[] add(Arg_0, Arg_1)
+  }
+
+  ENTRY main {
+    Arg_0.1 = s32[] parameter(0)
+    rd1 = s32[] all-reduce(Arg_0.1), to_apply=add.1
+    Arg_1.1 = s32[] parameter(1)
+    rd2 = s32[] all-reduce(Arg_1.1), to_apply=add.2
+    ROOT multiply.14 = s32[] multiply(rd1, rd2)
+  }
+  )";
+  auto computation_names = RunDeduplicatePass(text, /*expect_true=*/true);
+  EXPECT_EQ(computation_names.size(), 2);
+  for (auto name : computation_names) {
+    EXPECT_NE(name, "add.2");
+  }
+}
```

Comment:
can you add another test with a reduce and an allreduce. it would be great to test that after the dedup pass, the hlo graph won't change.

Reply:
Such a test is already here - see the DontDeduplicateReduceAllReduce test just above. Or have you had something else in mind?

(For completeness, the test above passed because of the IsCollectiveCalledComputation flag, now it passes because of the opcode comparison.)

Reply:
Yea similar to what @joker-eph  [suggested](https://github.com/openxla/xla/pull/8718#discussion_r1467118173) in the other PR. We'd need to be sure that this new mechanism is more reliable than using the IsCollectiveCalledComputation. I'm thinking something like this:
1. create a module with reduce and allreduce/reduceScatter
2. manually create a new computation(IsCollectiveCalledComputation is off by default)
3. replace the computation of allreduce with the one in 2 
4. run dedup pass and make sure HLO doesn't change.

Reply:
This is independent. It is a partial fix. The full fix will need many PRs:
- One of yours that you already have.
- this one.
- Others that add call to HLO and LLVM verifiers more frequently (this should have been catch up much earlier)
- And the refactoring of the attribute. Maybe we want them to be removed and inferred when needed or they should be cloned correctly.

@Tixxx what you describe shouldn't block this PR.

Reply:
Add the test. Is that what you had in mind?

Reply:
Yep, that looks good. Thanks.